packages/react/src/ReactLazy.js

```diff
@@ -49,44 +49,62 @@ export type LazyComponent<T, P> = {
 };
 
 function lazyInitializer<T>(payload: Payload<T>): T {
+  function onError(error) {
+    if (payload._status === Pending) {
+      // Transition to the next state.
+      const rejected: RejectedPayload = (payload: any);
+      rejected._status = Rejected;
+      rejected._result = error;
+    }
+  }
+
   if (payload._status === Uninitialized) {
     const ctor = payload._result;
     const thenable = ctor();
     // Transition to the next state.
     const pending: PendingPayload = (payload: any);
     pending._status = Pending;
     pending._result = thenable;
-    thenable.then(
-      moduleObject => {
-        if (payload._status === Pending) {
-          const defaultExport = moduleObject.default;
-          if (__DEV__) {
-            if (defaultExport === undefined) {
-              console.error(
-                'lazy: Expected the result of a dynamic import() call. ' +
-                  'Instead received: %s\n\nYour code should look like: \n  ' +
-                  // Break up imports to avoid accidentally parsing them as dependencies.
-                  'const MyComponent = lazy(() => imp' +
-                  "ort('./MyComponent'))",
-                moduleObject,
-              );
-            }
+    thenable.then(moduleObject => {
+      if (payload._status === Pending) {
+        let defaultExport;
+        if (__DEV__) {
+          if (moduleObject === undefined) {
+            console.error(
+              'lazy: Expected the result of a dynamic import() call. ' +
+                'Instead received: %s\n\nYour code should look like: \n  ' +
+                // Break up imports to avoid accidentally parsing them as dependencies.
+                'const MyComponent = lazy(() => imp' +
+                "ort('./MyComponent'))\n\n" +
+                'Did you accidentally put curly braces around the import?',
+              moduleObject,
+            );
           }
-          // Transition to the next state.
-          const resolved: ResolvedPayload<T> = (payload: any);
-          resolved._status = Resolved;
-          resolved._result = defaultExport;
         }
-      },
-      error => {
-        if (payload._status === Pending) {
-          // Transition to the next state.
-          const rejected: RejectedPayload = (payload: any);
-          rejected._status = Rejected;
-          rejected._result = error;
+        try {
+          defaultExport = moduleObject.default;
+        } catch (e) {
+          onError(e);
+          return;
         }
-      },
-    );
+        if (__DEV__) {
+          if (defaultExport === undefined) {
```

Comment:
Why isn't this also a `Rejected` case?

Reply:
What do you mean? It gets Rejected later on throw the catch. This particular check is just to give a nicer message but there could be more reasons `.default` access throws (e.g. a getter that throws) so I wanted to separate out the DEV message for one particular case and the production check.

Reply:
Sorry, to clarify, if `moduleObject.default` doesn't throw, but returns `undefined` for some other reason (no default export) then we'd consider it `Resolved`. Is that "supported"?

Reply:
I think that's "supported" in the sense that it's going to throw in reconciler later when we try to render it. The reconciler knows what it can or cannot render. So we don't need to exhaustively check it here.

Whereas if it's loading itself that fails, we don't even get to the reconciler. Because we end up caching `undefined` instead of a Promise, and keep throw `undefined`.

Reply:
I guess...but it feels weird to not handle it here, where we know the cause.

Reply:
I wanted to avoid adding duplicate production checks in a hot path since we're going to check the same thing later. So the only places I'm adding checks to are the ones we would _not_ check later because they're be incorrectly flagged as Pending. This is more about fixing the state machine of Lazy than about fixing what can or cannot get rendered.